compiler/noirc_frontend/src/elaborator/statements.rs

```diff
@@ -151,8 +151,24 @@ impl Elaborator<'_> {
         let (lvalue, lvalue_type, mutable) = self.elaborate_lvalue(assign.lvalue);
 
         if !mutable {
-            let (name, location) = self.get_lvalue_name_and_location(&lvalue);
+            let (_, name, location) = self.get_lvalue_error_info(&lvalue);
             self.push_err(TypeCheckError::VariableMustBeMutable { name, location });
+        } else if let Some(lambda_context) = self.lambda_stack.last() {
+            // We must check whether the mutable variable we are attempting to assign
+            // comes from a lambda capture. All captures are immutable so we want to error
+            // if the user attempts to mutate a captured variable inside of a lambda without mutable references.
+            let capture_ids =
+                lambda_context.captures.iter().map(|var| var.ident.id).collect::<Vec<_>>();
+            let (id, name, location) = self.get_lvalue_error_info(&lvalue);
+            let typ = self.interner.definition_type(id);
+            for capture_id in capture_ids {
+                if capture_id == id && !typ.is_mutable_ref() {
+                    self.push_err(TypeCheckError::MutableCaptureWithoutRef {
+                        name: name.clone(),
+                        location,
+                    });
+                }
+            }
```

Comment:
I didn't try it but I thought of a way to avoid collecting the capture IDs into a Vec:
```suggestion
            let (id, name, location) = self.get_lvalue_error_info(&lvalue);
            let typ = self.interner.definition_type(id);
            if !typ.is_mutable_ref() && lambda_context.captures.iter().any(|var| var.ident.id == id)
            {
                let name = name.clone();
                self.push_err(TypeCheckError::MutableCaptureWithoutRef { name, location });
            }
```